File: 백준/Platinum/11003. 최솟값 찾기/최솟값 찾기.java
import java.io.*;
import java.util.Deque;
import java.util.LinkedList;
import java.util.StringTokenizer;

public class Main {
    static class Node {
        public int value;
        public int index;

        Node(int value, int index) {
            this.value = value;
            this.index = index;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int l = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        Deque<Node> deque = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            int cur = Integer.parseInt(st.nextToken());
            while (!deque.isEmpty() && deque.getLast().value > cur) deque.removeLast();
            deque.addLast(new Node(cur, i));
            if (deque.getFirst().index <= i - l) deque.removeFirst();
            bw.write(deque.getFirst().value + " ");
        }

        bw.flush();
        br.close();
        bw.close();
    }
}

File: 백준/Platinum/11003. 최솟값 찾기/최솟값 찾기.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    var st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val l = st.nextToken().toInt()

    st = StringTokenizer(br.readLine())
    val deque = ArrayDeque<IntArray>()
    for (i in 0 until n) {
        val cur = st.nextToken().toInt()
        while (deque.isNotEmpty() && deque.last()[0] > cur) deque.removeLast()
        deque.addLast(intArrayOf(cur, i))
        if (deque.first()[1] <= i - l) deque.removeFirst()
        bw.write(deque.first()[0].toString())
        if (i < n - 1) bw.write(" ")
    }

    bw.flush()
    br.close()
    bw.close()
}